packages/in_app_purchase/in_app_purchase_platform_interface/lib/src/in_app_purchase_platform_addition.dart

```diff
@@ -2,18 +2,34 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-// ignore: avoid_classes_with_only_static_members
+import 'package:in_app_purchase_platform_interface/in_app_purchase_platform_interface.dart';
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+
 /// The interface that platform implementations must implement when they want to
 /// provide platform specific in_app_purchase features.
-abstract class InAppPurchasePlatformAddition {
+///
+/// Platform implementations should extend this class rather than implement it as `in_app_purchase`
+/// does not consider newly added methods to be breaking changes. Extending this class
+/// (using `extends`) ensures that the subclass will get the default implementation, while
+/// platform implementations that `implements` this interface will be broken by newly added
+/// [InAppPurchasePlatformAddition] methods.
+abstract class InAppPurchasePlatformAddition extends PlatformInterface {
```

Comment:
@stuartmorgan 

I'm making the `InAppPurchasePlatformAddition` also a `PlatformInterface`.
We didn't talk about this offline but I think this makes sense. Please share your thoughts.

Reply:
What's the advantage of doing this? It is not at all clear to me that if we added new methods to `PlatformInterface` they would be applicable to these addition classes.

Reply:
This is mainly to take advantage of using the token verification of the `PlatformInterface`.
I couldn't think of a scenario where we need to add some methods in the "PlatformAddition" that will not be applicable to all addition classes. 

Also, Now thinking about the usage of other "PlatformInterface"s, I'm not confident with `InAppPurchasePlatformAddition` being a "PlatformInterface". Maybe for this plugin i'll just duplicate the token verification code instead of extending `PlatformInterface`.

WDYT?

Reply:
Isn't the `verifyToken` meant to enforce users `extend` the class instead of `implement`ing it?

Is this a concern for the `InAppPurchasePlatformAddition` class?

I don't think we will need to be afraid of the `InAppPurchasePlatformAddition` class to have any methods. Meaning I don't expect any methods to be added in the future that would break implementing classes.

Reply:
We do need some sort of enforcement to make sure the addition is not a type of `InAppPurchasePlatform`.

Make sure an addition class to extend `InAppPurchasePlatformAddition` is one of the solution.

Reply:
Check, understood

Reply:
Extending happens to be the first easy solution I can think of; If there's an easier solution, we absolutely should adopt it!

Reply:
> We do need some sort of enforcement to make sure the addition is not a type of `InAppPurchasePlatform`.

Why do we need to enforce that? The concern is purely around pubspec dependency cycles with endorsement IIRC, in which case it's the pubspec cycle, not the classes, that are the issue. This seems like a very indirect way of trying to enforce a lack of cycles (and is both not sufficient to prevent a cycle, and not necessary in the case of an in-package implementation if this is something we adopt as a general solution).

Reply:
If the addition is a type of `InAppPurchasePlatform`, the plugin user would easily make a mistake to access the `InAppPurchasePlatform` object (e.g. InAppPurchasePlatformAndroid) and call the unified API there.

Reply:
Sorry, I'm not following. Can you give a snippet of what that mistake would look like in code?

Reply:
Let's say we allow an `InAppPurchasePlatform` to also be a type of `InAppPurchasePlatformAddition`, the implementer could potentially create a class like this:

```dart
class InAppPurchasePlatformGooglePlay extends InAppPurchasePlatform implements InAppPurchasePlatformAddition {
   
    @override
    Future<void> buyConsumable(...){...}

    Future<void> someAdditionMethod(...){...}
}
```

Then the user could potentially use this class to do something like:

```dart
InAppPurchasePlatformGooglePlay googlePlayAddition = InAppPurchaseConnection.instance.getPlatformAddition<InAppPurchasePlatformGooglePlay>();
googlePlayAddition.buyConsumable(...);
```

In this case, the user is accessing the platform interface API directly, which we want to avoid.

Reply:
I don't think we need to go out of our way to forbid that construction. It's a bad idea, but only in the same way that, say, making a bunch of implementation details of a `InAppPurchaseGoogleplayAdditions` public instead of private would be. I.e., it's the implementor of the addition class designing their own extension API badly.

What I wanted to avoid was the design where we *forced* extensions to use that, so that every extension would be required to have that poor design. If we make it so that it's easy to do the right thing, and model the right thing in our plugins, that should be enough.